tools/make_patch.c

```diff
@@ -0,0 +1,437 @@
+		if (name_len > sym_name_len) {
+			continue;
+		}
+		const char *sym_name = symbol->name;
+		if (name[0] == '.') {
+			// If `name` is a local label, compare it to the local part of `symbol->name`
+			sym_name += sym_name_len - name_len;
+		}
+		if (!strcmp(sym_name, name)) {
+			return symbol;
+		}
+	}
+	error_exit("Error: Unknown symbol: \"%s\"\n", name);
+}
+
+const struct Symbol *symbol_find_cat(const struct Symbol *symbols, const char *prefix, const char *suffix) {
+	char *sym_name = xmalloc(strlen(prefix) + strlen(suffix) + 1);
+	sprintf(sym_name, "%s%s", prefix, suffix);
+	const struct Symbol *symbol = symbol_find(symbols, sym_name);
+	free(sym_name);
+	return symbol;
+}
+
+int parse_number(const char *input, int base) {
+	char *endptr;
+	int n = strtol(input, &endptr, base);
+	if (endptr == input || *endptr || n < 0) {
+		error_exit("Error: Cannot parse number: \"%s\"", input);
+	}
+	return n;
+}
+
+void parse_symbol_value(char *input, int *restrict bank, int *restrict address) {
+	char *colon = strchr(input, ':');
+	if (!colon) {
+		error_exit("Error: Cannot parse bank+address: \"%s\"", input);
+	}
+	*colon++ = '\0';
+	*bank = parse_number(input, 16);
+	*address = parse_number(colon, 16);
+}
+
+void parse_symbols(const char *filename, struct Symbol **symbols) {
+	FILE *file = xfopen(filename, 'r');
+	struct Buffer *buffer = buffer_create(1);
+
+	enum { SYM_PRE, SYM_VALUE, SYM_SPACE, SYM_NAME } state = SYM_PRE;
+	int bank = 0;
+	int address = 0;
+
+	for (;;) {
+		int c = getc(file);
+		if (c == EOF || c == '\n' || c == '\r' || c == ';' || (state == SYM_NAME && (c == ' ' || c == '\t'))) {
+			if (state == SYM_NAME) {
+				// The symbol name has ended; append the buffered symbol
+				buffer_append(buffer, &(char []){'\0'});
+				symbol_append(symbols, buffer->data, bank, address);
+			}
+			// Skip to the next line, ignoring anything after the symbol value and name
+			state = SYM_PRE;
+			for (; c != EOF && c != '\n' && c != '\r'; c = getc(file));
+			if (c == EOF) {
+				break;
+			}
+		} else if (c != ' ' && c != '\t') {
+			if (state == SYM_PRE || state == SYM_SPACE) {
+				// The symbol value or name has started; buffer its contents
+				if (++state == SYM_NAME) {
+					// The symbol name has started; parse the buffered value
+					buffer_append(buffer, &(char []){'\0'});
+					parse_symbol_value(buffer->data, &bank, &address);
+				}
+				buffer->size = 0;
+			}
+			buffer_append(buffer, &c);
+		} else if (state == SYM_VALUE) {
+			// The symbol value has ended; wait to see if a name comes after it
+			state = SYM_SPACE;
+		}
+	}
+
+	fclose(file);
+	buffer_free(buffer);
+}
+
+int parse_arg_value(const char *arg, bool absolute, const struct Symbol *symbols, const char *patch_name) {
+	static const char *comparisons[] = {"==", ">", "<", ">=", "<=", "!=", "||"};
+	for (unsigned int i = 0; i < sizeof(comparisons) / sizeof(*comparisons); i++) {
+		if (!strcmp(arg, comparisons[i])) {
+			return i == 6 ? 0x11 : i; // "||" is 0x11
```

Comment:
It's very likely that other operators exist (like `&&` and maybe even bit-ops, maybe in other versions of the emulator), so either pad it out with NULL, make a n array of indexes, or do `i >= 6 ? i + 11 : i` for now.

Reply:
There are no other operators being used in any of the pokemon patch files. All the conditionals are already && by default. || Is only used when a conditiontype = 11 in other games. Some other operators MAY exist in the emulator and never used, but I'm less concerned with that right now.